src/LinuxTracing/LinuxTracingIntegrationTest.cpp

```diff
@@ -0,0 +1,300 @@
+  int perf_event_paranoid = 2;
+  bool atoi_succeeded = absl::SimpleAtoi(content, &perf_event_paranoid);
+  CHECK(atoi_succeeded);
+  return perf_event_paranoid;
+}
+
+[[nodiscard]] bool CheckIsPerfEventParanoidAtMost(int max_perf_event_paranoid) {
+  if (IsRunningAsRoot()) {
+    return true;
+  }
+
+  int perf_event_paranoid = ReadPerfEventParanoid();
+  if (perf_event_paranoid <= max_perf_event_paranoid) {
+    return true;
+  }
+
+  ERROR("Root or max perf_event_paranoid %d (actual is %d) required for this test",
+        max_perf_event_paranoid, perf_event_paranoid);
+  return false;
+}
+
+class ChildProcess {
+ public:
+  explicit ChildProcess(const std::function<int()>& child_main) {
+    std::array<int, 2> parent_to_child_pipe{};
+    CHECK(pipe(parent_to_child_pipe.data()) == 0);
+
+    std::array<int, 2> child_to_parent_pipe{};
+    CHECK(pipe(child_to_parent_pipe.data()) == 0);
+
+    pid_t child_pid = fork();
+    CHECK(child_pid >= 0);
+    if (child_pid > 0) {
+      // Parent.
+      child_pid_ = child_pid;
+
+      // Close unused ends of the pipes.
+      CHECK(close(parent_to_child_pipe[0]) == 0);
+      CHECK(close(child_to_parent_pipe[1]) == 0);
+
+      reading_fd_ = child_to_parent_pipe[0];
+      writing_fd_ = parent_to_child_pipe[1];
+
+    } else {
+      // Child.
+
+      // Close unused ends of the pipes.
+      CHECK(close(parent_to_child_pipe[1]) == 0);
+      CHECK(close(child_to_parent_pipe[0]) == 0);
+
+      // Redirect reading end of parent_to_child_pipe to stdin and close the pipe's original fd.
+      CHECK(close(STDIN_FILENO) == 0);
+      CHECK(dup2(parent_to_child_pipe[0], STDIN_FILENO) == STDIN_FILENO);
+      CHECK(close(parent_to_child_pipe[0]) == 0);
+
+      // Redirect writing end of child_to_parent_pipe to stdout and close the pipe's original fd.
+      CHECK(close(STDOUT_FILENO) == 0);
+      CHECK(dup2(child_to_parent_pipe[1], STDOUT_FILENO) == STDOUT_FILENO);
+      CHECK(close(child_to_parent_pipe[1]) == 0);
+
+      // Run the child and exit.
+      exit(child_main());
+    }
+  }
+
+  ~ChildProcess() {
+    CHECK(close(reading_fd_) == 0);
+    CHECK(close(writing_fd_) == 0);
+
+    CHECK(waitpid(child_pid_, nullptr, 0) == child_pid_);
+  }
+
+  [[nodiscard]] pid_t GetChildPid() const { return child_pid_; }
+
+  void WriteLine(std::string_view str) {
+    std::string string_with_newline = std::string{str}.append("\n");
+    CHECK(write(writing_fd_, string_with_newline.c_str(), string_with_newline.length()) ==
+          static_cast<ssize_t>(string_with_newline.length()));
+  }
+
+  [[nodiscard]] std::string ReadLine() {
+    std::string str;
+    while (true) {
+      char c;
+      CHECK(read(reading_fd_, &c, sizeof(c)) == sizeof(c));
+      if (c == '\n' || c == '\0') break;
+      str.push_back(c);
+    }
+    return str;
+  }
+
+ private:
+  pid_t child_pid_ = -1;
+  int reading_fd_ = -1;
+  int writing_fd_ = -1;
+};
+
+class BufferTracerListener : public TracerListener {
```

Comment:
Maybe BufferedTracerListener?

Reply:
I think that "buffered" gives the wrong message, I just want to say that this `TracerListener` simply *has* a buffer/vector/list where events are stored. Maybe "buffer" is not ideal? I could also just call this `SimpleTracerListener` if that's less confusing.

Reply:
Okay, then let's keep BufferTracerListener.